dali/pipeline/operators/displacement/warp_affine_params.h

```diff
@@ -0,0 +1,146 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_DISPLACEMENT_WARP_AFFINE_PARAMS_H_
+#define DALI_PIPELINE_OPERATORS_DISPLACEMENT_WARP_AFFINE_PARAMS_H_
+
+#include <vector>
+#include <sstream>
+#include "dali/pipeline/operators/operator.h"
+#include "dali/kernels/imgproc/warp/affine.h"
+#include "dali/kernels/imgproc/warp/mapping_traits.h"
+#include "dali/pipeline/operators/displacement/warp_param_provider.h"
+#include "dali/kernels/tensor_shape_print.h"
+
+namespace dali {
+
+template <int spatial_ndim>
+using WarpAffineParams = kernels::warp::mapping_params_t<kernels::AffineMapping<spatial_ndim>>;
+
+template <typename Backend,
+          int spatial_ndim,
+          typename BorderType>
+class WarpAffineParamProvider
+: public WarpParamProvider<Backend, spatial_ndim, WarpAffineParams<spatial_ndim>, BorderType> {
+ protected:
+  using MappingParams = WarpAffineParams<spatial_ndim>;
+  using Base = WarpParamProvider<Backend, spatial_ndim, MappingParams, BorderType>;
+  using Workspace = typename Base::Workspace;
+  using Base::ws_;
+  using Base::spec_;
+  using Base::params_gpu_;
+  using Base::params_cpu_;
+  using Base::num_samples_;
+
+  void SetParams() override {
+    if (spec_->NumRegularInput() >= 2) {
+      if (ws_->template InputIsType<GPUBackend>(1)) {
+        UseInputAsParams(ws_->template InputRef<GPUBackend>(1));
+      } else {
+        UseInputAsParams(ws_->template InputRef<CPUBackend>(1));
+      }
+    } else {
+      std::vector<float> matrix = spec_->template GetArgument<std::vector<float>>("matrix");
+      if (matrix.empty()) {
+        DALI_FAIL("`matrix` argument must be provided when transforms are not passed"
+                  " as a regular input.");
+      }
+
+      DALI_ENFORCE(matrix.size() == spatial_ndim*(spatial_ndim+1),
+        "`matrix` parameter must have " + std::to_string(spatial_ndim*(spatial_ndim+1)) +
+        " elements");
+
+      MappingParams M;
+      int k = 0;
+      for (int i = 0; i < spatial_ndim; i++)
+        for (int j = 0; j < spatial_ndim+1; j++, k++)
+          M.transform(i, j) = matrix[k];
+
+      auto *params = this->AllocParams(kernels::AllocType::Host);
+      for (int i = 0; i < params_cpu_.shape[0]; i++)
+        params[i] = M;
+    }
+  }
+
+  template <typename InputType>
+  void CheckParamInput(const InputType &input) {
+    DALI_ENFORCE(input.type().id() == DALI_FLOAT);
+
+    decltype(auto) shape = input.shape();
```

Comment:
wondering, why not `auto shape =`?

Reply:
because TensorList returns a reference and I don't need a copy - but TensorVector returns a value :(